### ISSUE TITLE
Clearly inform end users of CLI version mismatches

### ISSUE THREAD

**Issue by supertopher** [cloudfoundry/cli#1020]
TLDR: We have had major production issues on PWS due to CLI version mismatches.  These problems often manifest with error messages that waste time for development teams and users alike.  Users need better messaging about their CLI being out of date when errors happen.

Problem Statement: CLI users need to be more actively notified that their CLI is at a potentially incompatible level with their currently targeted API endpoint. We believe that any operation that requires a handshake should communicate that the CLI may be responsible if the CLI version < the min_CLI_version of the /v2/info endpoint. Due to human nature, we feel that this message should be more actively indicated (perhaps red text) when it may be a cause of failure. We know this often falls in the RTFM category, but feel that given the velocity of updates on server and CLI users need this extra nudge.

![](http://i.giphy.com/3orif7DKcHLWR6GMIE.gif)

**Context**
During a loggregator upgrade we changed the location of an endpoint.  This was the first major breaking change and left a large portion of our userbase in the dark as to why they could no longer push apps or tail logs.

Since this incident we have had other similar incidents.  Currently the cli will warn you both when you `target` and when you `auth`.  However these are two endpoints that are unlikely to have breaking changes across versions.  Thus our user base will do what we all do and ignore warnings until something breaks.  

Furthermore, the warning as currently written is buried in three other lines of text confirming positive results and no breaking error for the user will occur until a few minutes and many more terminal lines have passed.  At that point our end users see error messages that give the strong impression that PWS is broken.  

This behavior is bad for business and user happiness.  We have a real need to communicate very clearly to a user that when a command fails; their CF CLI version is a potential problem.  We cannot assume that customers will figure it out eventually.  We have competition and usability should be something take pride in and excel at.

The CLI team is in the middle of a codebase overhaul I would appreciate some thought around this problem for our users as you refactor the CLI.  You are of course the subject matter experts here, but this problem is very real to our platform as a whole.  

Thank you for your time and interest.  

CC:
@ccwu

**Comment by dkoper** [comment]
Hi @supertopher,

Can you give more concrete examples of bad UX due to version conpatibilities (you mentioned "other similar incidents" and buried warnings)

An issue with the change of loggregator URLs is that we currently don't reread the /v2/info for every command and I'm not sure it's worth the overhead of doing that - how often do operators change that url vs how many additional api requests the cf CLI would be making per user per day to be able to detect that change?
I believe the current message (from the noaa library) suggests to contact the CF operator. I have asked their PM to improve the message (with no luck), not sure we should try to trap that message and replace it. Maybe it's better if operators notify their users of such rare changes?

I'm interested in hearing more details of all your incidents and go through them one by one as I agree bad UX is bad for everybody.

Cheers,
Dies Koper
CF CLI PM

Open to

**Comment by abbyachau** [comment]
Hey @supertopher thanks for creating this issue. We've taken some steps to help mitigate against the issue you've raised. We recently [announced](https://github.com/cloudfoundry/cli/wiki/Versioning-Policy#cf-cli-minimum-supported-version) end of support for versions of cf-release older than v251. We've also created a [story](https://www.pivotaltracker.com/story/show/157229238) to warn users that their CF CLI version is no longer supported. I'm closing this issue but please feel free to reopen or create a new issue if you have any feedback, or if you have further suggestions. Happy to speak if you have any concerns. Thanks.
